### ISSUE TITLE
Transformer layer: attention + FFN with architecture dispatch

### ISSUE THREAD

**Issue by anibjoshi** [strata-ai-labs/strata-inference#12]
### Context
This is the core of the inference engine. A single transformer layer consists of:
1. Pre-normalization (LayerNorm or RMSNorm)
2. Multi-head attention (bidirectional or causal, with learned or RoPE positions)
3. Residual connection
4. Post-attention normalization
5. Feed-forward network (GELU or SwiGLU)
6. Residual connection

The implementation must be parameterized by ModelConfig to handle BERT, Gemma, and LLaMA from the same code path.

### Pseudocode
```rust
fn transformer_layer(
    input: &DeviceTensor,
    layer: &LayerWeights,
    config: &ModelConfig,
    backend: &dyn ComputeBackend,
    pos_offset: usize,  // for RoPE
) -> DeviceTensor {
    // 1. Pre-norm
    let normed = match config.norm_type {
        NormType::LayerNorm => backend.layer_norm(&input, &layer.attn_norm_w, &layer.attn_norm_b, config.norm_eps),
        NormType::RMSNorm => backend.rms_norm(&input, &layer.attn_norm_w, config.norm_eps),
    };

    // 2. QKV projection
    let q = backend.matmul(&normed, &layer.wq);
    let k = backend.matmul(&normed, &layer.wk);
    let v = backend.matmul(&normed, &layer.wv);

    // 3. Position encoding
    if config.position_type == PositionType::RoPE {
        backend.rope(&mut q, &mut k, pos_offset, config.rope_freq_base, config.head_dim);
    }

    // 4. Multi-head attention
    let attn_out = multi_head_attention(q, k, v, config, backend);

    // 5. Output projection + residual
    let projected = backend.matmul(&attn_out, &layer.wo);
    let residual1 = backend.add_tensor(&input, &projected);

    // 6. FFN pre-norm
    let normed2 = normalize(&residual1, &layer.ffn_norm_w, config, backend);

    // 7. FFN
    let ffn_out = match config.activation {
        Activation::GELU => {
            let up = backend.matmul(&normed2, &layer.ffn_up);
            let activated = backend.gelu(&up);
            backend.matmul(&activated, &layer.ffn_down)
        }
        Activation::SwiGLU => {
            let gate = backend.matmul(&normed2, &layer.ffn_gate);
            let up = backend.matmul(&normed2, &layer.ffn_up);
            let activated = backend.swiglu(&gate, &up);
            backend.matmul(&activated, &layer.ffn_down)
        }
    };

    // 8. Residual
    backend.add_tensor(&residual1, &ffn_out)
}
```

### LayerWeights Structure
```rust
pub struct LayerWeights {
    // Attention
    pub attn_norm_w: DeviceTensor,
    pub attn_norm_b: Option<DeviceTensor>,  // None for RMSNorm
    pub wq: DeviceTensor,
    pub wk: DeviceTensor,
    pub wv: DeviceTensor,
    pub wo: DeviceTensor,
    // FFN
    pub ffn_norm_w: DeviceTensor,
    pub ffn_norm_b: Option<DeviceTensor>,
    pub ffn_up: DeviceTensor,
    pub ffn_gate: Option<DeviceTensor>,  // None for non-gated (BERT)
    pub ffn_down: DeviceTensor,
}
```

### Tensor Name Mapping
GGUF tensor names follow patterns like:
- `blk.{i}.attn_q.weight` → wq for layer i
- `blk.{i}.attn_k.weight` → wk
- `blk.{i}.attn_v.weight` → wv
- `blk.{i}.attn_output.weight` → wo
- `blk.{i}.ffn_up.weight` → ffn_up
- `blk.{i}.ffn_gate.weight` → ffn_gate
- `blk.{i}.ffn_down.weight` → ffn_down
- `blk.{i}.attn_norm.weight` → attn_norm_w

### Acceptance Criteria
- [ ] `TransformerLayer` struct with config-driven forward pass
- [ ] Support both LayerNorm and RMSNorm normalization
- [ ] Support both learned positions and RoPE
- [ ] Support both GELU FFN and SwiGLU gated FFN
- [ ] Support both bidirectional and causal attention masks
- [ ] Multi-head attention with configurable num_heads and head_dim
- [ ] Grouped Query Attention (GQA) when num_kv_heads < num_heads
- [ ] Load layer weights from GGUF tensor names
- [ ] Test BERT-style layer vs reference outputs
- [ ] Test Gemma-style layer vs reference outputs

### Reference
- llama.cpp: `src/models/bert.cpp` (BERT builder)
- llama.cpp: `src/models/llama.cpp` (LLaMA builder with template<bool embed>)
- llama.cpp: `src/llama-graph.cpp` lines 925-1120 (build_norm, build_ffn)

### Estimated LOC: ~1,000-1,200

**Comment by anibjoshi** [comment]
## Critical Gaps from llama.cpp Evaluation

After comparing our implementation against llama.cpp, here are the critical items for the transformer layer:

### 1. Attention Logit Softcapping (CRITICAL for Gemma3)

Gemma3 applies softcapping to attention logits before softmax:

```rust
// After computing attention scores, before softmax:
if config.attn_logit_softcap > 0.0 {
    let cap = config.attn_logit_softcap;
    for score in attn_scores.iter_mut() {
        *score = cap * (*score / cap).tanh();
    }
}
```

This needs a new backend operation `softcap(tensor, cap_value)` or can be done inline. Without it, Gemma3 attention scores diverge from reference and produce incorrect embeddings. llama.cpp reads this from `{arch}.attn_logit_softcapping` metadata.

**Action:** Add `softcap` to `ComputeBackend` trait or implement inline in the attention computation.

### 2. Embedding Scaling (CRITICAL for Gemma)

Gemma models scale token embeddings by `sqrt(hidden_size)` immediately after lookup:

```rust
let embeddings = backend.embedding_lookup(&token_embd, &token_ids);
if config.embedding_scale != 1.0 {
    embeddings = backend.scale(&embeddings, config.embedding_scale);
}
```

llama.cpp applies this in the model forward pass. Without it, all downstream activations have the wrong magnitude.

**Action:** Apply scaling after embedding lookup in the forward pass, reading the scale factor from `ModelConfig`.

### 3. Partial RoPE (HIGH — correctness for some models)

Our current `rope()` applies rotation to ALL dimensions of each attention head. llama.cpp supports partial RoPE via `{arch}.rope.dimension_count`:

```rust
// Current (applies to all dims):
for i in 0..half_dim { ... }

// Correct (applies to first rope_dim/2 pairs, leaves rest unchanged):
let n_rot = config.rope_dim / 2;  // may be < half_dim
for i in 0..n_rot { ... }
// dims beyond n_rot are untouched
```

For EmbeddingGemma-300M, `rope_dim` likely equals `head_dim`, but we must read the metadata key to be sure.

**Action:** Add `rope_dim` parameter to the `rope()` backend call, defaulting to `head_dim`.

### 4. GQA K/V Head Repetition (CRITICAL when head_count_kv < head_count)

When using Grouped Query Attention, K and V tensors have fewer heads than Q. Before computing attention scores, K/V must be repeated:

```rust
fn repeat_kv(kv: &DeviceTensor, n_rep: usize) -> DeviceTensor {
    // kv shape: [seq_len, n_kv_heads, head_dim]
    // output:   [seq_len, n_heads, head_dim]
    // Each KV head is repeated n_rep = n_heads / n_kv_heads times
}
```

For EmbeddingGemma-300M: `head_count=16`, `head_count_kv` needs to be checked — if < 16, GQA is active.

**Action:** Implement `repeat_kv` helper. Check EmbeddingGemma-300M's `head_count_kv` value.

### 5. Attention Scale Factor

llama.cpp uses a configurable attention scale instead of always `1/sqrt(d_k)`:

```rust
let scale = config.attn_scale.unwrap_or(1.0 / (config.head_dim as f32).sqrt());
attn_scores = backend.scale(&attn_scores, scale);
```

Some models set `{arch}.attention.scale` explicitly. Default should be `1/sqrt(head_dim)`.

### 6. Updated Pseudocode for Attention Block

```rust
fn multi_head_attention(q, k, v, config, backend) -> DeviceTensor {
    // Reshape Q: [seq, hidden] → [seq, n_heads, head_dim]
    // Reshape K: [seq, hidden] → [seq, n_kv_heads, head_dim]  
    // Reshape V: [seq, hidden] → [seq, n_kv_heads, head_dim]
    
    // Apply RoPE (partial: only first rope_dim dimensions)
    backend.rope(&q, pos, config.rope_freq_base, config.head_dim, config.rope_dim);
    backend.rope(&k, pos, config.rope_freq_base, config.head_dim, config.rope_dim);
    
    // GQA: repeat K/V if needed
    if config.num_kv_heads < config.num_heads {
        k = repeat_kv(&k, config.num_heads / config.num_kv_heads);
        v = repeat_kv(&v, config.num_heads / config.num_kv_heads);
    }
    
    // Attention scores: Q @ K^T
    let scores = backend.matmul_transpose(&q, &k);
    let scores = backend.scale(&scores, attn_scale);
    
    // Softcap (Gemma3 only)
    if config.attn_logit_softcap > 0.0 {
        scores = softcap(&scores, config.attn_logit_softcap);
    }
    
    // Mask (causal or bidirectional based on config.causal)
    if config.causal {
        scores = backend.apply_causal_mask(&scores, seq_len);
    }
    
    // Softmax + weighted sum
    let weights = backend.softmax(&scores);
    let out = backend.matmul(&weights, &v);
    
    // Reshape back: [seq, n_heads, head_dim] → [seq, hidden]
    backend.matmul(&out, &layer.wo)
}
```

### 7. Backend Trait Additions Needed

| New Operation | Signature | Used By |
|---------------|-----------|---------|
| `softcap` | `(tensor, cap: f32) -> DeviceTensor` | Gemma3 attention |
| `repeat_kv` | `(tensor, n_rep: usize) -> DeviceTensor` | GQA models |

Or these can be implemented inline in the transformer layer code rather than as backend ops.

### Summary: Priority Order

1. **Embedding scaling** — Simple `scale()` call, blocks all downstream correctness
2. **Attention logit softcapping** — Required for Gemma3 attention correctness
3. **GQA K/V repetition** — Required if `head_count_kv < head_count`
4. **Partial RoPE** — Read `rope.dimension_count`, pass to `rope()` call
5. **Configurable attention scale** — Read from metadata, default to `1/sqrt(d_k)`

**Comment by anibjoshi** [comment]
Completed in PR #18 (M4: transformer layer).
